pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -92,6 +101,45 @@ type GraphBuilder struct {
 	absentOwnerCache *UIDCache
 }
 
+type stopMux struct {
```

Comment:
this `stopMux` thing does not make the code easier to understand IMO.  Just inline the mux logic into `monitorSingle.Run`.

It might even be much simpler to have one Go routine waiting on the global lock and shutting down all monitors in a loop. Any special reason that you want to have this logic per monitor?